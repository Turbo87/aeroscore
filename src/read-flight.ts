import Point from './geo/point';

import fs = require('fs');
import IGCParser = require('igc-parser');

export interface Fix {
  time: number;
  coordinate: Point;
  valid: boolean;
  altitude: number | null;
}

export function readFlightFromString(str: string): Fix[] {
  return IGCParser.parse(str).fixes.map(fix => ({
    time: fix.timestamp,
    coordinate: [fix.longitude, fix.latitude] as Point,
    valid: fix.valid,
    altitude: fix.gpsAltitude,
  }));
}

export function readFlight(path: string): Fix[] {
  return readFlightFromString(fs.readFileSync(path, 'utf8'));
}
