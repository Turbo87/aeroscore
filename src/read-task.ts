import bearing from '@turf/bearing';
import fs = require('fs');

import Point from './geo/point';
import {Cylinder, Keyhole, Line} from './task/shapes';
import Task from './task/task';
import {Turnpoint} from './turnpoint';
import {fraction} from './utils/angles';
import {read, XCSoarLocation} from './xcsoar';

export function readTaskFromString(str: string): Task {
  let task = read(str);

  let points = task.points.map((point, i) => {
    let location = convertLocation(point.waypoint.location);

    if (point.observation_zone.type === 'Cylinder') {
      return new Cylinder(location, point.observation_zone.radius!);
    }

    let direction;
    if (i === 0) {
      let locNext = convertLocation(task.points[i + 1].waypoint.location);
      direction = bearing(location, locNext);

    } else if (i === task.points.length - 1) {
      let locPrev = convertLocation(task.points[i - 1].waypoint.location);
      direction = bearing(locPrev, location);

    } else {
      let locPrev = convertLocation(task.points[i - 1].waypoint.location);
      let locNext = convertLocation(task.points[i + 1].waypoint.location);

      let bearingtoPrev = bearing(location, locPrev);
      let bearingToNext = bearing(location, locNext);

      let bisector = fraction(bearingtoPrev, bearingToNext);

      direction = bisector - 90;
    }

    if (point.observation_zone.type === 'Line') {
      return new Line(location, point.observation_zone.length!, direction);
    }

    if (point.observation_zone.type === 'Keyhole') {
      return new Keyhole(location, direction - 90);
    }

    throw new Error(`Unknown zone type: ${point.observation_zone.type}`);
  }).map(oz => new Turnpoint(oz));

  return new Task(points, {
    isAAT: task.type === 'AAT',
    aatMinTime: task.aat_min_time,
  });
}

export function readTask(path: string): Task {
  let file = fs.readFileSync(path, 'utf8');
  return readTaskFromString(file);
}

function convertLocation(loc: XCSoarLocation): Point {
  return [loc.longitude, loc.latitude];
}
