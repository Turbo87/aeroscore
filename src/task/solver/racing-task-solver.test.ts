import fs = require('fs');

import {readFlight} from '../../read-flight';
import {readTask} from '../../read-task';
import Task from '../task';
import RacingTaskSolver from './racing-task-solver';

const FIXTURES_PATH = `${__dirname}/../../../fixtures`;

const FIXTURES = [
  ['2017-07-17-lev', '2017-lev.csv', null],
  ['2017-07-17-lev', '2017-lev.csv', '2017-07-17T16:39:00Z'],
];

describe('RacingTaskSolver', () => {
  for (let [fixtureName, filterFilename, maxTime] of FIXTURES) {
    let testName = fixtureName!;
    if (maxTime !== null) {
      testName += ` until ${maxTime}`;
    }

    let maxT = Date.parse(maxTime!);

    test(testName, () => {
      let task = readTask(`${FIXTURES_PATH}/${fixtureName}.tsk`);
      let handicaps = readCSV(`${FIXTURES_PATH}/${filterFilename}`);

      let lines = findFlights(`${FIXTURES_PATH}/${fixtureName}/`)
        .map(({ callsign, flight }) => {
          let solver = new RacingTaskSolver(task);
          for (let fix of flight) {
            if (fix.time > maxT) break;

            solver.update(fix);
          }

          let result = solver.result;

          let handicap = handicaps[callsign.toUpperCase()];
          let handicapFactor = 100 / handicap;

          let { distance, speed } = result;
          if (distance !== undefined) distance *= handicapFactor;
          if (speed !== undefined) speed *= handicapFactor;

          return { callsign, handicap, result, distance, speed };
        })
        .sort(compareResults)
        .map((result: any) => {
          let distance = result.result.distance !== undefined ? `${(result.result.distance / 1000).toFixed(1)} km` : '';
          let speed = result.result.speed !== undefined ? `${(result.result.speed).toFixed(2)} km/h` : '';

          return [
            result.callsign,
            result.handicap.toString().padStart(3),
            distance.padStart(8),
            speed.padStart(11),
          ].join('\t  ');
        })
        .join('\n');

      expect(`\n${lines}\n`).toMatchSnapshot();
    });
  }

  describe('with task "2017-07-17-lev.tsk"', () => {
    let task: Task;
    let solver: RacingTaskSolver;

    beforeEach(() => {
      task = readTask(`${FIXTURES_PATH}/2017-07-17-lev.tsk`);
      solver = new RacingTaskSolver(task);
    });

    it('returns a result', () => {
      let flight = readFlight(`${FIXTURES_PATH}/2017-07-17-lev/IGP_77hg7sd1.IGC`);
      solver.consume(flight);
      expect(solver.result).toMatchSnapshot();
    });

    it('returns an intermediate result', () => {
      let flight = readFlight(`${FIXTURES_PATH}/2017-07-17-lev/IGP_77hg7sd1.IGC`);
      let part1 = flight.slice(0, 1500);
      let part2 = flight.slice(1500);
      solver.consume(part1);
      expect(solver.result).toMatchSnapshot();
      solver.consume(part2);
      expect(solver.result).toMatchSnapshot();
    });

    it('can handle outlandings', () => {
      let flight = readFlight(`${FIXTURES_PATH}/2017-07-17-lev/ZG_77hv6ci1.igc`);
      solver.consume(flight);
      expect(solver.result).toMatchSnapshot();
    });
  });
});

function findFlights(folderPath: string) {
  return fs.readdirSync(folderPath)
    .filter(filename => (/\.igc$/i).test(filename))
    .filter(filename => filename.match(/^(.{1,3})_/))
    .map(filename => {
      let callsign = filename.match(/^(.{1,3})_/)![1];
      let flight = readFlight(`${folderPath}/${filename}`);
      return { filename, callsign, flight };
    });
}

function compareResults(a: any, b: any) {
  if (a.speed !== undefined && b.speed !== undefined)
    return b.speed - a.speed;

  if (a.speed !== undefined && b.speed === undefined)
    return -1;

  if (a.speed === undefined && b.speed !== undefined)
    return 1;

  return b.distance - a.distance;
}

function readCSV(path: string) {
  let lines = fs.readFileSync(path, 'utf8').split('\n');
  lines.shift();

  let handicaps = Object.create(null);
  lines.map(line => line.trim().split(',')).forEach(([id, _, cn, type, handicap]) => {
    if (id) {
      handicaps[cn] = parseInt(handicap, 10);
    }
  });
  return handicaps;
}
