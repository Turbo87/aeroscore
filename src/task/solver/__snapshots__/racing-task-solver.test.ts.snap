// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" can handle outlandings 1`] = `
Object {
  "completed": false,
  "distance": 94373.825,
  "path": Array [
    StartEvent {
      "point": lon 6.9179704 lat 51.1276447,
      "time": 2017-07-17T14:06:54.931Z,
      "type": "start",
    },
    TurnEvent {
      "num": 1,
      "point": lon 6.2812555 lat 50.9890756,
      "time": 2017-07-17T15:42:59.779Z,
      "type": "turn",
    },
    TurnEvent {
      "num": 2,
      "point": lon 6.1937554 lat 51.1722784,
      "time": 2017-07-17T16:21:04.919Z,
      "type": "turn",
    },
  ],
  "speed": undefined,
  "time": undefined,
}
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" returns a result 1`] = `
Object {
  "completed": true,
  "distance": 136420.85,
  "path": Array [
    StartEvent {
      "point": lon 6.9313498 lat 51.0847306,
      "time": 2017-07-17T14:50:56.825Z,
      "type": "start",
    },
    TurnEvent {
      "num": 1,
      "point": lon 6.2829181 lat 50.9964654,
      "time": 2017-07-17T15:24:26.380Z,
      "type": "turn",
    },
    TurnEvent {
      "num": 2,
      "point": lon 6.1941852 lat 51.1723459,
      "time": 2017-07-17T15:51:51.555Z,
      "type": "turn",
    },
    TurnEvent {
      "num": 3,
      "point": lon 6.6075025 lat 50.9694879,
      "time": 2017-07-17T16:20:04.143Z,
      "type": "turn",
    },
    FinishEvent {
      "point": lon 7.0084235 lat 51.0985126,
      "time": 2017-07-17T16:39:10.837Z,
      "type": "finish",
    },
  ],
  "speed": 75.625831,
  "time": 6494.0122,
}
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" returns an intermediate result 1`] = `
Object {
  "completed": false,
  "distance": 52735.160,
  "path": Array [
    StartEvent {
      "point": lon 6.9147855 lat 51.1378604,
      "time": 2017-07-17T14:17:08.833Z,
      "type": "start",
    },
    TurnEvent {
      "num": 1,
      "point": lon 6.2829181 lat 50.9964654,
      "time": 2017-07-17T15:24:26.380Z,
      "type": "turn",
    },
  ],
  "speed": undefined,
  "time": undefined,
}
`;

exports[`RacingTaskSolver with task "2017-07-17-lev.tsk" returns an intermediate result 2`] = `
Object {
  "completed": true,
  "distance": 136420.85,
  "path": Array [
    StartEvent {
      "point": lon 6.9313498 lat 51.0847306,
      "time": 2017-07-17T14:50:56.825Z,
      "type": "start",
    },
    TurnEvent {
      "num": 1,
      "point": lon 6.2829181 lat 50.9964654,
      "time": 2017-07-17T15:24:26.380Z,
      "type": "turn",
    },
    TurnEvent {
      "num": 2,
      "point": lon 6.1941852 lat 51.1723459,
      "time": 2017-07-17T15:51:51.555Z,
      "type": "turn",
    },
    TurnEvent {
      "num": 3,
      "point": lon 6.6075025 lat 50.9694879,
      "time": 2017-07-17T16:20:04.143Z,
      "type": "turn",
    },
    FinishEvent {
      "point": lon 7.0084235 lat 51.0985126,
      "time": 2017-07-17T16:39:10.837Z,
      "type": "finish",
    },
  ],
  "speed": 75.625831,
  "time": 6494.0122,
}
`;
